### DISCUSSION TITLE
Renovate decides not to reuse branch when it looks un-modified

### DISCUSSION THREAD

**Discussion by osigida** [renovatebot/renovate#13011] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

29.30.1

### Please select which platform you are using if self-hosting.

Bitbucket Cloud (bitbucket.org)

### Describe the bug
Renovate stuck with some of the branches. 
It cannot rebase them on top of the master automatically nor with `rebase!` prefix.
However for some other branches it works well and can perform merges.

Form the log it looks like that Renovate decides to not reuse the branch when there are no changes in it, instead it recreates branch but later it is rejected.

related code:
https://github.com/renovatebot/renovate/blob/06d8c270433206eecba4f60635f669cd8c5393c1/lib/workers/branch/reuse.ts#L55-L63

the code is a bit confusing, see my question below:

```
    if (await isBranchStale(branchName)) {
      logger.debug(`Branch is stale and needs rebasing`);
      // We can rebase the branch only if no PR or PR can be rebased
      if (await isBranchModified(branchName)) {
        // TODO: Warn here so that it appears in PR body (#9720)
        logger.debug('Cannot rebase branch as it has been modified');
        return { reuseExistingBranch: true, isModified: true };
      }
    // Question: at this point isBranchModified returned FALSE, ie branch was NOT modified, why then we move on with reuseExistingBranch = false?
    // should it be like follow instead?
    //return { reuseExistingBranch: true, isModified: false };
      return { reuseExistingBranch: false };  
```

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Branch already exists (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: getBranchPr(renovate/spring-cloud) (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: findPr(renovate/spring-cloud, undefined, open) (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: getPrList() (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Found PR #472 (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: isBranchStale=true (repository=repository_name_here, branch=renovate/spring-cloud)
       "isStale": true,
       "currentBranch": "master",
       "currentBranchSha": "f53d52354a35fc83360770115db3f7ad6342ddde"
DEBUG: Branch is stale and needs rebasing (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Using reuseExistingBranch: false (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: manager.getUpdatedPackageFiles() reuseExistinbranch=false (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Updating org.springframework.cloud:spring-cloud-starter-contract-stub-runner in web/build.gradle (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Updating org.springframework.cloud:spring-cloud-starter-bootstrap in bootstrap/build.gradle (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Updated 2 package files (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: No updated lock files in branch (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: 2 file(s) to commit (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Committing files to branch renovate/spring-cloud (repository=repository_name_here, branch=renovate/spring-cloud)
DEBUG: Setting git author name (repository=repository_name_here, branch=renovate/spring-cloud)
       "gitAuthorName": "Renovate Bot"
DEBUG: Setting git author email (repository=repository_name_here, branch=renovate/spring-cloud)
       "gitAuthorEmail": "renovate@everon.io"
DEBUG: git commit (repository=repository_name_here, branch=renovate/spring-cloud)
       "deletedFiles": [],
       "ignoredFiles": [],
       "result": {
         "author": null,
         "branch": "renovate/spring-cloud",
         "commit": "26a5348",
         "root": false,
         "summary": {"changes": 2, "insertions": 2, "deletions": 2}
       }
 INFO: Branch update was rejected because local copy is not up-to-date. (repository=repository_name_here, branch=renovate/spring-cloud)
 ```
</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
Could you expand/rephrase your bug description so that it is more understandable?

**Comment by osigida** [comment]
@rarkins is it better now?

**Comment by rarkins** [comment]
I don't see anything wrong in the logs, apart from the last part where Bitbucket Cloud rejects the push. Is there a line before that which you think indicates a mistake?

**Comment by osigida** [comment]
indeed, but why local copy become not up-to-date, when there were no changes?

**Comment by osigida** [comment]
@rarkins I added few more lines to debug to make this particular flow clear
https://github.com/renovatebot/renovate/pull/13009

once it is merged and released, I guess it will be easier to understand what is going on

**Comment by rarkins** [comment]
I don't think those will help.

Based on the log message, it's this logic being triggered:

```
    if (err.message.includes('[rejected] (stale info)')) {
      logger.info(
        'Branch update was rejected because local copy is not up-to-date.'
      );
      return null;
    }
```

It's most likely something like this: https://stackoverflow.com/questions/56191415/why-is-git-push-force-with-lease-failing-with-rejected-stale-info-even

BTW I suggest you try again with persistRepoData=false if you're using it

**Comment by osigida** [comment]
Indeed we use `persistRepoData: true,` will try it.

However that sounds like a huge performance issue, fetch + rebase should be kinda enough... 

I would expect Renovate to be able to deal with such kinda issues, or be able to recover somehow. It looks like it misses `git rebase` somewhere in the flow...

**Comment by rarkins** [comment]
My best guess is that the solution lies in the link I gave above.

**Comment by osigida** [comment]
`Adding the --prune option to my initial git pull (before doing my rebase) corrects this problem.` ? 
maybe, but that means Renovate could do that as well 

for me it looks like Renovate does not merge branches/PRs it opened itself, and which have no changes form outside

**Comment by rarkins** [comment]
> `Adding the --prune option to my initial git pull (before doing my rebase) corrects this problem.` ? maybe, but that means Renovate could do that as well

Yes. Something for you to test locally?

> for me it looks like Renovate does not merge branches/PRs it opened itself, and which have no changes form outside

I don't know what you mean by this

**Comment by osigida** [comment]
@rarkins I have no issues to rebase a branch manually, but that I a bit silly to rebase Renovate's branches. 
I would expect it rebase itself, however it does not work like that. 

the flow looks like that 
1) renovate open a PR (this dependency has a history of updates, so there were PR's and branches)
2) master has changed, one of the developers merged something 
3) renovate tries to update the PR and fails to rebase, to merge, with the same error  "Branch update was rejected because local copy is not up-to-date." 

For some of the branches Renovate still able to rebase, but not for all of them.

So I think Renovate should be able to recover and rebase stale branches, and I don't understand why it cannot do it for my cases or how to help it to continue to merge updates automatically.

**Comment by osigida** [comment, marked as answer]
`persistRepoData=false` helped (on the second run, looks like it did clean up in the end of the first run)
means there is an issue in the way it fetch and rebase change requests

**Reply by rarkins** [reply]
I have transferred this to a discussion. It seems like your repo got out of sync as part of persistRepoData=true. There's most likely nothing wrong with Renovate's branch rebase logic because it does many thousands of them every day

**Reply by Morl99** [reply]
Hello, I wanted to chime in, as we experience very similar problems regularly. As far as we know, this started around 3 months ago and we are chasing the problems ever since. We are running Renovate self hosted in version 28.1.0 (currently updating to latest as per your comment, that the debug should be improved) on a self-hosted GitLab. 

Is there anything that we could do to better debug the issue?  I will disable `persistRepoData` now, since this is most likely the culprit, but I think it would be great, if we can help fixing the underlying issue with that option.
